### PR TITLE
Laboratory work #4, Anna Azimkova - 22FPL1

### PR DESCRIPTION
**Pull request by puglyann**

### CODE REVIEW

**Comment by puglyann** [comment]
how can i delete lab3😭

File: lab_4_fill_words_by_ngrams/main.py [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+
+        Raises:
+            ValueError: In case of inappropriate type input argument or if input argument is empty.
+        """
+        if not isinstance(text, str) or not text:
+            raise ValueError
+
+        punctuation = '!.?'
+        for element in text.lower():
+            if not element.isalnum():
```
Comment (marina-kaz) [inline lab_4_fill_words_by_ngrams/main.py:36]: digits are supposed to be removed

File: lab_4_fill_words_by_ngrams/main.py [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+        for word in decoded_corpus:
+            if word != self._end_of_word_token:
+                if not resulting_text:
+                    resulting_text += word.capitalize()
+                elif resulting_text[-1] == '.':
+                    resulting_text += ' ' + word.capitalize()
+                else:
+                    resulting_text += ' ' + word
+            else:
+                resulting_text += '.'
```
Comment (marina-kaz) [inline lab_4_fill_words_by_ngrams/main.py:90]: It is better to put shorter code blocks first for better readability

**Review by marina-kaz — COMMENTED** [review]
Nice start, go on!

File: lab_4_fill_words_by_ngrams/main.py (hunk 1 above)
Comment (puglyann) [reply lab_4_fill_words_by_ngrams/main.py]: rewrote it

File: lab_4_fill_words_by_ngrams/main.py [hunk 3]
```diff
@@ -98,6 +143,42 @@ def run(self, seq_len: int, prompt: str) -> str:  # type: ignore
...
+
+            sorted_tokens = sorted(list(tokens.items()), key=lambda pair: (float(pair[1]), -pair[0]), reverse=True)
+            probability = 0
+            for index, token_prob in enumerate(sorted_tokens):
+                probability += token_prob[1]
+                if probability >= self._p_value:
+                    random_token = random.choice(sorted_tokens[:index + 1])
+                    encoded += (random_token[0],)
+                    break
+            else:
```
Comment (marina-kaz) [inline lab_4_fill_words_by_ngrams/main.py:172]: else what?

File: lab_4_fill_words_by_ngrams/main.py [hunk 4]
```diff
@@ -98,6 +143,42 @@ def run(self, seq_len: int, prompt: str) -> str:  # type: ignore
...
+            tokens = self._model.generate_next_token(encoded)
+            if tokens is None:
+                raise ValueError
+
+            if not tokens:
+                break
+
+            sorted_tokens = sorted(list(tokens.items()), key=lambda pair: (float(pair[1]), -pair[0]), reverse=True)
+            probability = 0
+            for index, token_prob in enumerate(sorted_tokens):
```
Comment (marina-kaz) [inline lab_4_fill_words_by_ngrams/main.py:166]: consider iterating with `index, (token, token_prob)`, this should be more readable

File: lab_4_fill_words_by_ngrams/main.py [hunk 5]
```diff
@@ -98,6 +143,42 @@ def run(self, seq_len: int, prompt: str) -> str:  # type: ignore
...
+                and isinstance(seq_len, int) and seq_len > 0
+        ):
+            raise ValueError
+
+        encoded = self._word_processor.encode(prompt)
+
+        if not encoded:
+            raise ValueError
+
+        while seq_len >= 1:
```
Comment (marina-kaz) [inline lab_4_fill_words_by_ngrams/main.py:156]: Is there any variation in terms of how many steps there will be? If not, `for` should be used

**Review by marina-kaz — COMMENTED** [review]
Ok, keep going!

**Comment by puglyann** [comment]
wow 100 hundred commits
it's anniversary🎉🎉🎉

File: lab_4_fill_words_by_ngrams/main.py [hunk 6]
```diff
@@ -296,6 +460,16 @@ def assess_exam(self, answers: dict[str, str]) -> float:  # type: ignore
         Raises:
             ValueError: In case of inappropriate type input argument or if input argument is empty.
         """
+        if not isinstance(answers, dict) or not answers:
+            raise ValueError
+
+        correct_answers = 0
+
+        for key in self._questions_and_answers.keys():
+            if answers[key[0]] == self._questions_and_answers[key]:
```
Comment (marina-kaz) [inline lab_4_fill_words_by_ngrams/main.py:469]: consider iterating over items then if you need value as well

File: lab_4_fill_words_by_ngrams/main.py [hunk 7]
```diff
@@ -273,6 +424,18 @@ def _load_from_json(self) -> dict[tuple[str, int], str]:  # type: ignore
...
+
+        with open(self._json_path, 'r', encoding='utf-8') as file:
+            q_as = json.load(file)
+
+        if not isinstance(q_as, list):
+            raise ValueError
+
+        self._questions_and_answers = {(q_a['question'], q_a['location']): q_a['answer']
+                                       for q_a in q_as}
+        return self._questions_and_answers
```
Comment (marina-kaz) [inline lab_4_fill_words_by_ngrams/main.py:438]: why both assigning attribute and returning value, i believe it should just be returned

**Review by marina-kaz — COMMENTED** [review]
Good!
There are some unaddressed comments left

**Comment by vitazaytseva** [comment]
Well done! Good job!

**Pull request closed without merging**